corehq/apps/sso/tests/test_authentication.py

```diff
@@ -0,0 +1,138 @@
+from django.test import TestCase
+
+from corehq.apps.domain.models import Domain
+from corehq.apps.sso.authentication import get_authenticated_sso_user
+from corehq.apps.sso.exceptions import SsoAuthenticationError
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+from corehq.apps.sso.tests import generator
+from corehq.apps.users.models import WebUser
+
+
+class TestGetAuthenticatedSsoUser(TestCase):
+    """
+    This ensures that get_authenticated_sso_user succeeds only when the
+    request has met all required criteria.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.account = generator.get_billing_account_for_idp()
+        cls.domain = Domain.get_or_create_with_name("vaultwax-001", is_active=True)
+
+        # this will be the user that's "logging in" with SAML2 via the SsoBackend
+        cls.user = WebUser.create(
+            cls.domain.name, 'b@vaultwax.com', 'testpwd', None, None
+        )
+        cls.idp = generator.create_idp('vaultwax', cls.account)
+        cls.idp.is_active = True
+        cls.idp.save()
+        AuthenticatedEmailDomain.objects.create(
+            email_domain='vaultwax.com',
+            identity_provider=cls.idp,
+        )
+
+        idp_vwx = generator.create_idp('vwx-link', cls.account)
+        AuthenticatedEmailDomain.objects.create(
+            email_domain='vwx.link',
+            identity_provider=idp_vwx,  # note which idp is mapped here
+        )
+
+    @classmethod
+    def tearDownClass(cls):
+        AuthenticatedEmailDomain.objects.all().delete()
+        IdentityProvider.objects.all().delete()
+        cls.user.delete(None)
+        cls.domain.delete()
+        cls.account.delete()
+        super().tearDownClass()
+
+    def test_authentication_error_if_idp_doesnt_exist(self):
+        """
+        get_authenticated_sso_user should throw an SsoAuthenticationError if
+        the Identity Provider slug passed to it points to a non-existent
+        Identity Provider.
+        """
+        with self.assertRaises(SsoAuthenticationError):
+            get_authenticated_sso_user(
+                self.user.username,
+                'doesnotexist'
+            )
+
+    def test_authentication_error_if_idp_not_active(self):
+        """
+        get_authenticated_sso_user should throw an SsoAuthenticationError if
+        the specified Identity Provider is not active.
+        """
+        self.idp.is_active = False
+        self.idp.save()
+        with self.assertRaises(SsoAuthenticationError):
+            get_authenticated_sso_user(
+                self.user.username,
+                self.idp.slug
+            )
+
+        self.idp.is_active = True
+        self.idp.save()
```

Comment:
This should be done with `self.addCleanup(...)` or `try`/`finally`. Otherwise if this test fails (before this code is executed) it will cause other tests to fail as well. It's easiest to do try/finally in a suggestion on github, but feel free to use `addCleanup` instead if you prefer that.
```suggestion
        try:
            with self.assertRaises(SsoAuthenticationError):
                get_authenticated_sso_user(
                    self.user.username,
                    self.idp.slug
                )
        finally:
            self.idp.is_active = True
            self.idp.save()
```